More detailed documentation for recipes (#2343)
NVIDIA/TransformerEngine: docs/features/low_precision_training/fp8_current_scaling/fp8_current_scaling.rst

* E4M3 -- consists of 1 sign bit, 4 exponent bits and 3 bits of mantissa. It can store values up to +/-448 and ``nan``.
* E5M2 -- consists of 1 sign bit, 5 exponent bits and 2 bits of mantissa. It can store values up to +/-57344, +/- ``inf`` and ``nan``. The tradeoff of the increased dynamic range is lower precision of the stored values.

.. raw:: html
   :file: img/fp8_formats.svg

*Figure 1: Structure of the floating point datatypes. All of the values shown (in FP16, BF16, FP8 E4M3 and FP8 E5M2) are the closest representations of value 0.3952.*

**E4M3 and E5M2 usage in training**

By default, Transformer Engine uses a hybrid approach:

* *Forward pass* - activations and weights require more precision, so E4M3 datatype is best used.
* *Backward pass* - gradients are less susceptible to precision loss but require higher dynamic range, so E5M2 datatype is preferred. 

The user can configure this behavior via the ``fp8_format`` parameter of the recipe.

Scaling factors
---------------

FP8's limited dynamic range is insufficient for many tensors. 

> To address this, scaling factors are used. In FP8 Current Scaling there is one **FP32** scale factor per tensor.

Reviewer: ```suggestion
To address this, values in the tensor are scaled. FP8 Current Scaling recipe uses one **FP32** scale factor per tensor.
```

The representation of a tensor element ``x`` in FP8 precision is given by:

.. code-block:: python

    x = x_fp8 * s

where

* ``x_fp8`` is the FP8 value (E4M3 or E5M2),
* ``s`` is a global **FP32** scaling factor applied to the entire tensor.

**FP8 Current Scaling quantization**

Let's look more closely at how quantization to FP8 with scaling factor is implemented in
the FP8 Current Scaling recipe.

.. raw:: html
   :file: img/fp8_scaling_concept.svg

> *Figure 3: Quantization to FP8 consists of amax computation, scaling to fit the FP8 range and casting to the respective FP8 format.*

Reviewer: Also, I don't like the "FP8 range min" in the picture - this is ambiguous, considering that we are mostly talking about the dynamic range in the text - in that context, it should be something like "- FP8 range max". Maybe we could do something like:
 - first picture: add "Original range" spanning all values
 - second picture: rather than having a single arrow show a trapezoid that compacts that original range to the FP8 range
 - third picture: no changes
